// belvo/lib/belvo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _http = _interopRequireDefault(require("./http"));

var _accounts = _interopRequireDefault(require("./accounts"));

var _balances = _interopRequireDefault(require("./balances"));

var _institutions = _interopRequireDefault(require("./institutions"));

var _invoices = _interopRequireDefault(require("./invoices"));

var _links = _interopRequireDefault(require("./links"));

var _owners = _interopRequireDefault(require("./owners"));

var _statements = _interopRequireDefault(require("./statements"));

var _incomes = _interopRequireDefault(require("./incomes"));

var _taxComplianceStatus = _interopRequireDefault(require("./taxComplianceStatus"));

var _investmentsPortfolios = _interopRequireDefault(require("./investmentsPortfolios"));

var _investmentsTransactions = _interopRequireDefault(require("./investmentsTransactions"));

var _recurring_expenses = _interopRequireDefault(require("./recurring_expenses"));

var _risk_insights = _interopRequireDefault(require("./risk_insights"));

var _taxReturns = _interopRequireDefault(require("./taxReturns"));

var _taxStatus = _interopRequireDefault(require("./taxStatus"));

var _taxRetentions = _interopRequireDefault(require("./taxRetentions"));

var _transactions = _interopRequireDefault(require("./transactions"));

var _widgetToken = _interopRequireDefault(require("./widgetToken"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Client = /*#__PURE__*/function () {
  function Client(secretKeyId, secretKeyPassword) {
    var url = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;

    _classCallCheck(this, Client);

    var belvoUrl = (0, _utils.urlResolver)(url || process.env.BELVO_API_URL);

    if (!belvoUrl) {
      throw new Error('You need to provide a URL or a valid environment.');
    }

    this.session = new _http["default"](belvoUrl);
    this.secretKeyId = secretKeyId;
    this.secretKeyPassword = secretKeyPassword;
  }

  _createClass(Client, [{
    key: "connect",
    value: function () {
      var _connect = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var login;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.session.login(this.secretKeyId, this.secretKeyPassword);

              case 2:
                login = _context.sent;

                if (login) {
                  _context.next = 5;
                  break;
                }

                throw new Error('Login failed.');

              case 5:
                this.institutions = new _institutions["default"](this.session);
                this.links = new _links["default"](this.session);
                this.accounts = new _accounts["default"](this.session);
                this.transactions = new _transactions["default"](this.session);
                this.owners = new _owners["default"](this.session);
                this.balances = new _balances["default"](this.session);
                this.invoices = new _invoices["default"](this.session);
                this.taxReturns = new _taxReturns["default"](this.session);
                this.taxComplianceStatus = new _taxComplianceStatus["default"](this.session);
                this.taxStatus = new _taxStatus["default"](this.session);
                this.statements = new _statements["default"](this.session);
                this.incomes = new _incomes["default"](this.session);
                this.recurringExpense = new _recurring_expenses["default"](this.session);
                this.riskInsights = new _risk_insights["default"](this.session);
                this.widgetToken = new _widgetToken["default"](this.session);
                this.investmentsPortfolios = new _investmentsPortfolios["default"](this.session);
                this.investmentsTransactions = new _investmentsTransactions["default"](this.session);
                this.taxRetentions = new _taxRetentions["default"](this.session);

              case 23:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function connect() {
        return _connect.apply(this, arguments);
      }

      return connect;
    }()
  }]);

  return Client;
}();

var _default = Client;
exports["default"] = _default;
